eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -140,264 +459,307 @@ proposalsAddVote voter vote gai (Proposals omap) =
+              , "> HardFork"
+              , show $ ps ^. pHierarchyL . pfhHardForkL . pHierarchyNTL
+              , "> Committee"
+              , show $ ps ^. pHierarchyL . pfhCommitteeL . pHierarchyNTL
+              , "> Constitution"
+              , show $ ps ^. pHierarchyL . pfhConstitutionL . pHierarchyNTL
+              ]
+            else mempty
+         )
+      <> ["----- Proposals End -----"]
 
-pgacPParamUpdateL ::
-  Lens' (PrevGovActionIdsChildren era) (Set (PrevGovActionId 'PParamUpdatePurpose (EraCrypto era)))
-pgacPParamUpdateL = lens pgacPParamUpdate $ \x y -> x {pgacPParamUpdate = y}
+-- | Get the sequence of `GovActionState`s
+proposalsActions ::
+  Proposals era ->
+  StrictSeq (GovActionState era)
+proposalsActions (Proposals omap _ _) = OMap.toStrictSeq omap
 
-pgacHardForkL ::
-  Lens' (PrevGovActionIdsChildren era) (Set (PrevGovActionId 'HardForkPurpose (EraCrypto era)))
-pgacHardForkL = lens pgacHardFork $ \x y -> x {pgacHardFork = y}
+-- | Get the sequence of `GovActionId`s
+proposalsIds ::
+  Proposals era ->
+  StrictSeq (GovActionId (EraCrypto era))
+proposalsIds (Proposals omap _ _) = OMap.toStrictSeqOKeys omap
 
-pgacCommitteeL ::
-  Lens' (PrevGovActionIdsChildren era) (Set (PrevGovActionId 'CommitteePurpose (EraCrypto era)))
-pgacCommitteeL = lens pgacCommittee $ \x y -> x {pgacCommittee = y}
+-- | Get the unordered map of `GovActionId`s and `GovActionState`s
+proposalsGovActionStates ::
+  Proposals era ->
+  Map (GovActionId (EraCrypto era)) (GovActionState era)
+proposalsGovActionStates (Proposals omap _ _) = OMap.toMap omap
 
-pgacConstitutionL ::
-  Lens' (PrevGovActionIdsChildren era) (Set (PrevGovActionId 'ConstitutionPurpose (EraCrypto era)))
-pgacConstitutionL = lens pgacConstitution $ \x y -> x {pgacConstitution = y}
+proposalsSize :: Proposals era -> Int
+proposalsSize (Proposals omap _ _) = OMap.size omap
 
-data PrevGovActionIdsChildren era = PrevGovActionIdsChildren
-  { pgacPParamUpdate :: !(Set (PrevGovActionId 'PParamUpdatePurpose (EraCrypto era)))
-  , pgacHardFork :: !(Set (PrevGovActionId 'HardForkPurpose (EraCrypto era)))
-  , pgacCommittee :: !(Set (PrevGovActionId 'CommitteePurpose (EraCrypto era)))
-  , pgacConstitution :: !(Set (PrevGovActionId 'ConstitutionPurpose (EraCrypto era)))
+proposalsLookupId ::
+  GovActionId (EraCrypto era) ->
+  Proposals era ->
+  Maybe (GovActionState era)
+proposalsLookupId gai (Proposals omap _ _) = OMap.lookup gai omap
+
+deriving instance Era era => Eq (PForest StrictMaybe era)
+deriving instance Era era => Ord (PForest StrictMaybe era)
+deriving instance Era era => Generic (PForest StrictMaybe era)
+deriving instance Era era => NoThunks (PForest StrictMaybe era)
+deriving instance Era era => NFData (PForest StrictMaybe era)
+deriving instance Era era => ToJSON (PForest StrictMaybe era)
+deriving instance Era era => Default (PForest StrictMaybe era)
+
+newtype PrevGovActionIds era = PrevGovActionIds
```

Comment:
I wonder if we shouldn't rename this as well, since we eradicated `PrevGovActionId` . Maybe `GovPurposeIds` ?

Reply:
The `Prev` prefix was originally intended to mean that these were enacted on the previous epoch boundary. Since there is no semantic change to this type, keeping it as is, one could argue, still does that job of conveying this meaning.

Reply:
Oh I see, that makes sense.